File: 0438-find-all-anagrams-in-a-string/0438-find-all-anagrams-in-a-string.js
/**
 * @param {string} s
 * @param {string} p
 * @return {number[]}
 */
var findAnagrams = function(s, p) {
  const count = new Array(26).fill(0);
  const a = 'a'.charCodeAt();
  const ans = [];
  let [l, r, t] = [0, 0, p.length];
  
  for (let ch of p) 
    count[ch.charCodeAt() - a]++;
  
  while (r < s.length) {
    if (count[s.charCodeAt(r++) - a]-- > 0) t--;
    if (!t) ans.push(l);
    
    while (r - l >= p.length) 
      if (++count[s.charCodeAt(l++) - a] > 0) t++;
  }
  
  return ans;
};

File: 0438-find-all-anagrams-in-a-string/0438-find-all-anagrams-in-a-string.py
class Solution:
    def findAnagrams(self, s: str, p: str) -> List[int]:
        count, ans = Counter(p), []
        n, a = len(p), ord('a')
        j, t = 0, n
        for i, ch in enumerate(s):
            if count[ch] > 0:
                t -= 1
            count[ch] -= 1
            if t == 0:
                ans.append(j)
            while i - j + 1 >= n:
                count[s[j]] += 1
                if count[s[j]] > 0:
                    t += 1
                j += 1
        return ans
                
                